drivers: net: cobs l2 network interface with async uart transport

mkschreder opened the pull request:
### Summary 
This PR adds a new way to run Zephyr IP networking over a simple TX/RX UART link using cobs for framing. Zephyr currently does not have a proper serial line network interface with multi instance support because the existing one uses uart pipe. 

The interface I'm adding in this PR uses predictable overhead COBS framing that is much better than slip and I'm also adding this as an interface of which you can have many instances. This interface is primarily meant for reliable board to board networking and I'm also adding throughput samples for both UDP and TCP so you can test how this works. 

**NOTE**: this is a big PR and still needs to be cleaned up. Also I'm still working on refactoring the samples and my goal is to make them bare yet robust. I also plan to update docs and describe in detail how this works.

- **A new COBS UART async network driver** that exposes a standard Zephyr network interface over UART.
- **A small COBS-based framing layer for serial links** so the UART link has clear packet boundaries.
- **New samples** to make bring-up and evaluation easy:
  - `samples/net/cobs_uart_async` (basic interface + net shell)
  - `samples/net/cobs_throughput_client` / `samples/net/cobs_throughput_server` (UDP echo-based throughput/latency)
  - `samples/net/cobs_throughput_single` and `samples/net/cobs_throughput_tcp` (single-board, cross-wired UART testing)
- **New tests** covering both “bytes on the wire” and “sockets on top”:
  - `tests/drivers/cobs_uart_async_loopback`
  - `tests/drivers/cobs_uart_async_sockets`
- **Documentation + devicetree binding** to make configuration straightforward.

### Why this is useful
- **Reliable serial networking**: enables IP connectivity on systems that only have a basic UART link available (TX/RX + GND), without requiring hardware flow control.
- **Easier evaluation and regression coverage**: provides ready-to-run samples for throughput testing and adds automated tests to keep the full stack (driver → network stack → BSD sockets) working over time.
- **Cleaner framing behavior**: COBS framing is both efficient and necessary in this case for packet based communication on a stream-based transport like UART.

### What’s included (high level)
- **Driver**: `drivers/net/cobs_uart_async.c` (+ Kconfig + build integration)
- **Serial framing support**: new COBS Serial L2 + net_pkt COBS helper APIs
- **Devicetree**: binding for declaring one or more COBS UART interfaces
- **Samples + tests**: end-to-end examples and validation coverage for real hardware and simulation scenarios

### Dependencies

- https://github.com/zephyrproject-rtos/zephyr/pull/101004: COBS library extensions.

### Testing
- Added unit test and a socket integration samples to validate end-to-end behavior over COBS framed packet link.

mkschreder committed:
lib: cobs: implement streaming cobs variant

Many transports deliver data in small, arbitrary chunks (for example,
UART RX buffers). In these cases, treating COBS encoding/decoding as a
single “whole buffer in / whole buffer out” operation is awkward.

This paves the way for being able to add highly efficient cobs-based
interfaces that operate on chunks instead of whole buffers. We avoid
creating a new buffer just to hold incomplete frame and instead maintain
state as we are receiving more data.

Encoding is a bit more complex since cobs requires lookahead when
encoding, but this is not a problem because we have access to full data
when we encode so we can do lookahead when needed.

Signed-off-by: Martin Schröder <info@swedishembedded.com>

jukkar commented on drivers/net/slip_uart_async.c:
```diff
@@ -0,0 +1,441 @@
+#define LOG_LEVEL CONFIG_SLIP_UART_ASYNC_LOG_LEVEL
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
Instead of this, you could just write
```
#include <zephyr/logging/log.h>
LOG_MODULE_REGISTER(slip_uart_async, CONFIG_SLIP_UART_ASYNC_LOG_LEVEL);
```
which is the style we are mainly using in zephyr.

jukkar commented on drivers/net/slip_uart_async.c:
I am not sure if it is a good idea to print non ascii chars here, it does not give much benefit and might cause some garbage to be printed.
Also this looks like a debug print than an info message. Usually there is no need to print this kind of messages unless you are debugging things.

mkschreder replied:
I think this is gone now.

jukkar commented on samples/net/slip_throughput_client/boards/native_sim.overlay:
```diff
@@ -0,0 +1,17 @@
+/*
```
About the samples, could you please place them in a separate commit. Also we could create `samples/net/slip` and place the related samples there.

jukkar commented on samples/net/slip_throughput_client/src/main.c:
```diff
@@ -0,0 +1,439 @@
+	/* Set receive timeout */
+	timeout.tv_sec = 1;
+	timeout.tv_usec = 0;
+	ret = zsock_setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
```
Please check the code and do not mix POSIX and network namespaced symbols. We recently merged a PR that adds a prefix to various network symbols in order to avoid conflicts with POSIX.
So here use `ZSOCK_SOL_SOCKET` and `ZSOCK_SO_RCVTIMEO`
If you want to check the usage of the correct APIs, you can set `CONFIG_NET_NAMESPACE_COMPAT_MODE=n` to disable the compatibility mode and make sure correct symbols are used.

mkschreder replied:
These samples will be cleaned up as I mentioned in the note in PR description.

jukkar commented on samples/net/slip_throughput_client/README.rst:
```diff
@@ -0,0 +1,156 @@
+- IPv6: fe80::2
+
+Default test parameters (can be overridden via Kconfig or shell):
+- Server IP: 192.168.7.1
```
Please change this

jukkar reviewed (commented):
Nice work! There is a lot of code and I will continue the review later, this was just a first glance of the PR.
Couple of thoughts about this:
* It might be confusing if this is called SLIP as it typically means certain framing for the packets as seen in https://en.wikipedia.org/wiki/Serial_Line_Internet_Protocol. Should we mention the COBS there or something like that?
* Would it be possible to create a Linux counter part of the protocol so that this could be used from qemu or even from native_sim? So similar way the SLIP can be used.

mkschreder commented on doc/connectivity/networking/api/slip_uart_async.rst:
Sure. I think I will update them to be ipv6 addresses though.

mkschreder commented on include/zephyr/data/cobs.h:
```diff
@@ -86,8 +86,109 @@ int cobs_encode(struct net_buf *src, struct net_buf *dst, uint32_t flags);
+ * Non-destructive: tracks read position without modifying source.
+ */
+struct cobs_encode_state {
+	struct net_buf *src_frag;   /* Current source fragment being read */
```
This will disappear once the cobs MR is merged.

mkschreder committed:
drivers: net: add cobs l2 async uart network interface support

Add a new way to run Zephyr IP networking over a simple TX/RX UART link
using cobs for framing. Zephyr currently does not have a proper serial
line network interface with multi instance support because the existing
one uses uart pipe.

mkschreder committed:
lib: cobs: refactor to support custom delimiter and consolidate

- Add back support for custom delimiter everywhere
- Consolidate cobs_encode/cobs_decode to call streaming api
- Add additional test cases for 0x00 block and fix that as well
- Remove implicit conversions to boolean

Signed-off-by: Martin Schröder <info@swedishembedded.com>
